parlai/agents/hugging_face/gpt2.py

```diff
@@ -0,0 +1,210 @@
+
+    def output(self, tensor):
+        """
+        Compute output logits.
+        """
+        # get only scores for labels
+        if self.text_lengths is not None:
+            total_length = max(self.text_lengths)
+            to_select = tensor.size(1) - total_length + 1
+            if to_select > 0:
+                # select only label scores
+                bsz = tensor.size(0)
+                new_tensors = []
+                for i in range(bsz):
+                    start = self.text_lengths[i] - 1
+                    end = start + to_select
+                    new_tensors.append(tensor[i, start:end, :].unsqueeze(0))
+                tensor = torch.cat(new_tensors, 0)
+
+        return self.lm_head(tensor)
+
+    def reorder_decoder_incremental_state(self, incremental_state, inds):
+        new_incr_state = []
+        for layer_past in incremental_state:
+            reordered_layer_past = [
+                layer_past[:, i].unsqueeze(1).clone().detach() for i in inds
+            ]
+            reordered_layer_past = torch.cat(reordered_layer_past, dim=1)
+            new_incr_state.append(reordered_layer_past)
+
+        return tuple(new_incr_state)
+
+    def decode_forced(self, encoder_states, ys):
+        """
+        Override to get rid of start token input.
+        """
+        seqlen = ys.size(1)
+        inputs = ys.narrow(1, 0, seqlen - 1)
+        latent, _ = self.decoder(inputs, encoder_states)
+        logits = self.output(latent)
+        _, preds = logits.max(dim=2)
+        return logits, preds
+
+    def forward(self, *xs, ys=None, prev_enc=None, maxlen=None, bsz=None):
+        model_input, text_lengths = xs
+        if ys is not None:
+            self.text_lengths = text_lengths
+        else:
+            self.text_lengths = None
+
+        return super().forward(
+            model_input, ys=ys, prev_enc=prev_enc, maxlen=maxlen, bsz=bsz
+        )
+
+
+############################################
+## Agent
+############################################
+
+
+class Gpt2Agent(TorchGeneratorAgent):
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        agent = argparser.add_argument_group('Gpt2 Args')
+        agent.add_argument(
+            '--gpt2-size',
+            type=str,
+            default='small',
+            choices=['small', 'medium', 'large', 'xl'],
+            help='Which size model to initialize.',
+        )
+        agent.add_argument(
+            '--add-special-tokens',
+            type='bool',
+            default=True,
+            help='Add special tokens (like PAD, etc.). If False, '
+            'Can only use with batch size 1.',
+        )
+        argparser.set_defaults(
+            text_truncate=768,
+            label_truncate=256,
+            dict_maxexs=0,  # skip building dictionary
+        )
+        super(Gpt2Agent, cls).add_cmdline_args(argparser)
+        warn_once('WARNING: this model is in beta and the API is ' 'subject to change.')
```

Comment:
nit: you can remove the extra `'` here since it looks like black combined to one line